### ISSUE TITLE
Error messages from launch condition not shown to user except through "View Log".

### ISSUE THREAD

**Issue by monty241** [oleg-shilo/wixsharp#156]
Not related to #150. On trunk.

When you have an error such as:

```
Info: Action start 12:06:39: LaunchConditions.
Error: Invantive Bridge Connectors for PowerBI requires Microsoft .NET Framework 462 package or greater to be installed. Please install a supported Microsoft .NET Framework version.
Info: Action ended 12:06:39: LaunchConditions. Return value 3.
Info: Action ended 12:06:39: INSTALL. Return value 3.
```

You will with PR #155 get a message "Interrupted" in the title. But the actual error is not shown such as a popup box:

> Invantive Bridge Connectors for PowerBI requires Microsoft .NET Framework 462 package or greater to be installed. Please install a supported Microsoft .NET Framework version.

So the user sees that the installation was interrupted, but the launch condition message is not shown. Unless he opens the log through 'view log' button.

This will typically either support calls or abandoning of a package app.

**Comment by monty241** [comment]
PR upcoming.

**Comment by monty241** [comment]
Change is include in PR #155 , commit 69fb036f6964014de0eff6fc5e24f82d0d069e5e

**Comment by monty241** [comment]
Result: 
![image](https://user-images.githubusercontent.com/7579444/30111605-a150cb26-930e-11e7-9144-3ce9823510cc.png)

**Comment by oleg-shilo** [comment]
This is a difficult one. No, the solution you are proposing is easy to implement but it cannot be accepted as a common solution. 

You probably noticed that the native MSI UI never shows the errors as an extra modal message box. One good reason against that is that modal dialogs are in general a controversial UX concept. Though I doubt MSI designers were concerned much about UX. :) 

The more important reason is that the session may have multiple errors and it's impossible to say which one should be reported as a failure cause. 

Implementing your proposal outside of MSI/WiX (e.g. in Wix#) is even more difficult as the error message in many cases will require localization. Even the message box title "Attention" will need to be translated into other languages and this cannot be done as MSI/WiX localization does not cover your message box but MSI dialogs only.

Thus the actual failure cause can only be discovered by reading the log. Of course MSI made it ridiculously difficult by not providing the direct access to the log file. This is something that Wix# addressed well.

Saying that, I understand your intention and made a change in the `IManagedUIShell` interface to provide the required infrastructure for achieving your objective. Now you have an option of overriding one of the dialogs where you can access the complete set of the detected errors. Now you can decide by yourself which error to show, how and when.

```C#
void ExitDialog_Load(object sender, System.EventArgs e)
{
    image.Image = MsiRuntime.Session.GetEmbeddedBitmap("WixUI_Bmp_Dialog");
    if (Shell.UserInterrupted)
    {
        title.Text = "[UserExitTitle]";
        description.Text = "[UserExitDescription1]";
        this.Localize();
    }
    else if (Shell.ErrorDetected)
    {
        title.Text = "[FatalErrorTitle]";
        description.Text = "[FatalErrorDescription1]";
        this.Localize();
    }

    ResetLayout();

    if (Shell.Errors.Any())
        MessageBox.Show(Shell.Errors.Last(), "Attention");
}
```

**Comment by monty241** [comment]
OK thanks for considering it carefully. I don't understand the concepts of msi so well but based upon the msdn page I thought to understand that they wanted each message to be shown in a message box because they were so easy to cast :-) 
This solution is also great and sufficient for our purposes. Have a great weekend!

**Comment by oleg-shilo** [comment]
Hi Guido,

I have an update to on the solution to this issue. 

I did some extra experiments and digging in the WiX/MSI internals. And in result I have changed my opinion about the best approach for reporting the LaunchCondition failure (e.g. downgrade detected) to the user. I am now convinced that your original proposal for popping up the message box from ProcessMessage handler is in fact a preferred approach. You were right on money from the start. :)

The reasons for this:
* Despite my initial believe, I found that MSI does not do comprehensive error analysis and simply treats the first detected error as a failure. 

* MSI also ignores all user triggered errors  (e.g. CA return error value) and logs them as "INFO". Of course it is a questionable approach but it lets MSI to be aggressive towards any errors found (simply because it will not be many). This means that these errors cannot be detected in the stream of multiple true "INFO" message. 

* MSI also does not bother with localization but shows raw error message with a default message box title. Again, simplistic but reasonably adequate approach.  

That is why I have incorporated the following routine from your PR:
```C#
default:
{
    if (messageType.IsOneOf(InstallMessage.Error, InstallMessage.Warning, InstallMessage.User))
    {
        MessageBox.Show(
            this.shellView,
            messageRecord.ToString(),
            "[ErrorDlg_Title]".LocalizeWith(MsiRuntime.Localize),
            (MessageBoxButtons)(int)buttons,
            (MessageBoxIcon)(int)icon,
            (MessageBoxDefaultButton)(int)defaultButton);
    }
    ... 
```

I will close your PR as its all major elements have been implemented (but simply outside of the standard git PR merge process).

Thank you.
